test/integration/delta_cds_integration_test.cc

```diff
@@ -0,0 +1,213 @@
+    //    Envoy used in the tests.
+    // 4) Bringing up the server usually entails waiting to ensure that any listeners specified in
+    //    the bootstrap config have come up, and registering them in a port map (see lookupPort()).
+    //    However, this test needs to defer all of that to later.
+    defer_listener_finalization_ = true;
+    HttpIntegrationTest::initialize();
+
+    // Create the regular (i.e. not an xDS server) upstream. We create it manually here after
+    // initialize() because finalize() expects all fake_upstreams_ to correspond to a static
+    // cluster in the bootstrap config - which we don't want since we're testing dynamic CDS!
+    fake_upstreams_.emplace_back(new FakeUpstream(0, FakeHttpConnection::Type::HTTP2, version_,
+                                                  timeSystem(), enable_half_close_));
+    fake_upstreams_[UpstreamIndex]->set_allow_unexpected_disconnects(false);
+
+    // Now that the upstream has been created, process Envoy's request to discover it.
+    // (First, we have to let Envoy establish its connection to the CDS server.)
+    AssertionResult result = // xds_connection_ is filled with the new FakeHttpConnection.
+        fake_upstreams_[0]->waitForHttpConnection(*dispatcher_, xds_connection_);
+    RELEASE_ASSERT(result, result.message());
+    result = xds_connection_->waitForNewStream(*dispatcher_, xds_stream_);
+    RELEASE_ASSERT(result, result.message());
+    xds_stream_->startGrpcStream();
+    fake_upstreams_[0]->set_allow_unexpected_disconnects(true);
+
+    EXPECT_TRUE(compareDeltaDiscoveryRequest(Config::TypeUrl::get().Cluster, {}, {}));
+    sendDeltaDiscoveryResponse<envoy::api::v2::Cluster>({buildCluster(ClusterName)}, {}, "1");
+    // We can continue the test once we're sure that Envoy's ClusterManager has made use of
+    // the DiscoveryResponse describing cluster_0 that we sent.
+    // 2 because the statically specified CDS server itself counts as a cluster.
+    test_server_->waitForGaugeGe("cluster_manager.active_clusters", 2);
+
+    // Wait for our statically specified listener to become ready, and register its port in the
+    // test framework's downstream listener port map.
+    test_server_->waitUntilListenersReady();
+    registerTestServerPorts({"http"});
+  }
+};
+
+INSTANTIATE_TEST_CASE_P(IpVersionsClientType, CdsIntegrationTest, GRPC_CLIENT_INTEGRATION_PARAMS);
+
+// 1) Envoy starts up with no static clusters (other than the CDS-over-gRPC server).
+// 2) Envoy is told of a cluster via CDS.
+// 3) We send Envoy a request, which we verify is properly proxied to and served by that cluster.
+// 4) Envoy is told that cluster is gone.
+// 5) We send Envoy a request, which should 503.
+// 6) Envoy is told that the cluster is back.
+// 7) We send Envoy a request, which we verify is properly proxied to and served by that cluster.
+TEST_P(CdsIntegrationTest, CdsClusterUpDownUp) {
```

Comment:
Should we verify delta behavior, e.g. send 5 clusters, remove 3, add 4 more?

Reply:
Sure, a more interesting test would be good; I will add that. Although, other than the general fact of testing things beyond "zero" and "one" being a good thing, I think it won't really test the delta behavior "more." The current test, using DeltaDiscoveryRequest/Response, adds/removes/adds a cluster. You can make the Envoy's state follow the same path with an appropriate sequence of state-of-the-world updates. Similarly, you can get the same Envoy states of e.g. {1,2,3,4,5}, {1,5}, {1,2,3,5,6,7} with state-of-the-world updates.

(...which makes me think, is it really being tested enough? Maybe the delta-as-opposed-to-SotW testing needs to be in unit tests, if this integration test can only manage "does it get the job done").

Reply:
Test added.